### DISCUSSION TITLE
Access Token Renewal Issue on ESP32 (M5Stack)

### DISCUSSION THREAD

**Discussion by hitoyoshi-y** [mobizt/ESP-Google-Sheet-Client#42] [Q&A]
Hello,

I am encountering an issue while trying to use the append function on my ESP32 board (M5Stack). The following error occurs when attempting to refresh the access token for the 4th time:

Token expiration time: Token info: type = OAuth2.0 access token, status = on exchange request
> ERROR.mUpdateEngine: Error writing to basic client.
> WARN.mRunUntil: Terminating with write error.
> ERROR.mConnectSSL: Failed to initialize the SSL layer.
> ERROR.mConnectSSL: Unknown error code.
Token info: type = OAuth2.0 access token, status = error
Token error: code: -4, message: connection lost
1970/01/01 09:00:00
current_time: 2024/12/04 20:42:42
Free Memory: 228788 bytes

After this error, if I restart the microcontroller, it works fine for the first 3 access token renewals, but the issue reoccurs on the 4th attempt.

Could you please share any insights into what might be causing this issue and how to resolve it?

Thank you for your time and assistance.

Best regards,

**Comment by mobizt** [comment]
There is no such issue in re-authentication (token refreshment).

You can force token to be refreshed whenever you want with `refreshToken()`.
https://github.com/mobizt/ESP-Google-Sheet-Client?tab=readme-ov-file#force-the-token-to-expire-immediately-and-refresh

You can test by refreshing the token repeatedly at some interval e.g. every 1 minute.

Or using `setPrerefreshSeconds(60)` for testing.
https://github.com/mobizt/ESP-Google-Sheet-Client?tab=readme-ov-file#set-the-seconds-to-refesh-auth-token-before-it-expires


Actually, the error is because network (WiF) was disconnected.

> Token error: code: -4, message: connection lost

**Reply by hitoyoshi-y** [reply]
Thank you for your response!!

I reviewed the advice you provided.
First, I tried using GSheet.refreshToken() to refresh the token every 10 seconds while testing GSheet.values.append().
The result was the same: append() succeeded up to the third attempt, but the same error occurred on the fourth attempt.

Regarding the second point about network connectivity, I checked the WiFi.status() value when the error occurred, and it showed 3 (WL_CONNECTED), so the network connection seems to be fine.

Could this issue be related to the configuration of the Google service account?

**Reply by mobizt** [reply]
The time 10 sec is too short. You should test with longer time interval.

Here the code that works with token refreshment.

I recommend use reset instead.

I test this before post here.

This code runs for many years without any problems.

```cpp

#include <ESP_Google_Sheet_Client.h>

#define PROJECT_ID "........"

#define CLIENT_EMAIL "........"

const char PRIVATE_KEY[] PROGMEM = "........";

#define USER_EMAIL "........"

#define WIFI_SSID "........"
#define WIFI_PASSWORD "........"

unsigned long ms = 0;

int counter = 0;

void tokenStatusCallback(TokenInfo info);

void setup()
{

    Serial.begin(115200);
    Serial.println();
    Serial.println();

    GSheet.printf("ESP Google Sheet Client v%s\n\n", ESP_GOOGLE_SHEET_CLIENT_VERSION);

    WiFi.setAutoReconnect(true);

    WiFi.begin(WIFI_SSID, WIFI_PASSWORD);

    Serial.print("Connecting to Wi-Fi");
    unsigned long ms = millis();
    while (WiFi.status() != WL_CONNECTED)
    {
        Serial.print(".");
        delay(300);
    }
    Serial.println();
    Serial.print("Connected with IP: ");
    Serial.println(WiFi.localIP());
    Serial.println();

    GSheet.setTokenCallback(tokenStatusCallback);

    GSheet.begin(CLIENT_EMAIL, PROJECT_ID, PRIVATE_KEY);
}

void loop()
{

    bool ready = GSheet.ready();

    if (ready && millis() - ms > 8000)
    {
        ms = millis();

        FirebaseJson response;

        Serial.println("\nAppend spreadsheet values...");
        Serial.println("----------------------------");

        FirebaseJson valueRange;

        valueRange.add("majorDimension", "COLUMNS");
        String path1 = "values/[0]/[";
        for (int i = 0; i < 5; i++)
        {
            String path = path1;
            path += i;
            path += "]";
            valueRange.set(path, random(100, 300));
        }

        bool success = GSheet.values.append(&response, "........", "Sheet1!B3:C5", &valueRange, "USER_ENTERED", "INSERT_ROWS");
        if (success)
            response.toString(Serial, true);
        else
            Serial.println(GSheet.errorReason());
        Serial.println();

        Serial.println(ESP.getFreeHeap());

        counter++;

        if (counter == 2)
        {
            GSheet.reset();
            counter = 0;
        }
    }
}

void tokenStatusCallback(TokenInfo info)
{
    if (info.status == token_status_error)
    {
        GSheet.printf("Token info: type = %s, status = %s\n", GSheet.getTokenType(info).c_str(), GSheet.getTokenStatus(info).c_str());
        GSheet.printf("Token error: %s\n", GSheet.getTokenError(info).c_str());
    }
    else
    {
        GSheet.printf("Token info: type = %s, status = %s\n", GSheet.getTokenType(info).c_str(), GSheet.getTokenStatus(info).c_str());
    }
}

```

**Reply by mobizt** [reply]
> Could this issue be related to the configuration of the Google service account?

No.

**Reply by mobizt** [reply]
Google will limit the google API request per minute for DDoS attack prevention.  

If you send too often requests to Google API server (e.g. token refreshment), Google will deny the request and you have to wait for few minutes to start sending the request (refresh) again.

Your error information you posted can be the consequence of some situations that should be investigated further e.g. out of memory, unstable network connection and power supply issue.
